src/sqswatcher/sqswatcher.py

```diff
@@ -206,68 +232,145 @@ def _requeue_message(queue, message):
+        parsed_events = _parse_messages_helper(
+            messages, sqs_config, queue_type=QueueType.health, table=table, queue=health_queue,
+        )
+        return _resolve_events_hostname_collision(event_pool=parsed_events[EventType.HEALTH])
+    except Exception as e:
+        logging.error("Failed to parse messages from %s with exception: %s", health_queue.queue_name, e)
+    return []
+
+
+def _parse_messages_helper(messages, sqs_config, queue_type, table=None, queue=None):
+    parsed_events = {
+        EventType.ADD: [],
+        EventType.REMOVE: [],
+        EventType.HEALTH: [],
+    }
+    event_name_to_processing_function = {
+        EventType.HEALTH: lambda: _process_scheduled_maintenance_event(
+            sqs_config.stack_name, sqs_config.region, sqs_config.proxy_config, message_attrs, message, table, queue
+        ),
+        EventType.ADD: lambda: _process_compute_ready_event(
+            sqs_config.region, sqs_config.proxy_config, message_attrs, message, table
+        ),
+        EventType.REMOVE: lambda: _process_instance_terminate_event(message_attrs, message, table, queue),
+    }
+    event_name_to_event_type = {
+        "parallelcluster:EC2_SCHEDULED_EVENT": EventType.HEALTH,
+        "parallelcluster:COMPUTE_READY": EventType.ADD,
+        "autoscaling:EC2_INSTANCE_TERMINATE": EventType.REMOVE,
+    }
+    for message in messages:
+        # CW event rule can only send messages to SQS with "" enclosing the message.
+        # However, when message is re-queued it is sent without ""
+        message_text = (
+            json.loads(json.loads(message.body)) if str(message.body).startswith('"') else json.loads(message.body)
+        )
+        message_attrs = json.loads(message_text.get("Message"))
+        event_name = message_attrs.get("Event")
+        if not event_name:
+            log.warning("Unable to read message. Deleting.")
+            message.delete()
+            continue
+        event_type = event_name_to_event_type.get(event_name, None)
+
+        if event_type in SUPPORTED_EVENTTYPE_FOR_QUEUETYPE[queue_type]:
+            event = event_name_to_processing_function.get(event_type)()
+            if event:
+                parsed_events[event_type].append(event)
+        else:
+            log.info("Unsupported event type %s for %s queue. Discarding message.", event_type, queue_type)
+            message.delete()
+
+    return parsed_events
+
+
+def _process_scheduled_maintenance_event(stack_name, region, proxy_config, message_attrs, message, table, queue):
+    try:
+        instance_id = message_attrs.get("EC2InstanceId")
+        instances_in_cluster = get_cluster_instance_info(
+            stack_name, region, proxy_config, instance_ids=[instance_id], include_master=False
+        )
+        if instance_id in instances_in_cluster:
+            hostname = _retrieve_hostname_from_ddb(instance_id, table)
+            log.info("Relevant EC2 scheduled event for instance:%s in ASG.", instance_id)
```

Comment:
minor: remove ASG at the end? also might want to add a space after `:`

Reply:
Done